esp-wifi/src/compat/misc.rs

```diff
@@ -20,8 +21,26 @@ unsafe extern "C" fn fclose(stream: *const ()) -> i32 {
     todo!("fclose {:?}", stream);
 }
 
-// not available in ROM on ESP32-S2
-#[cfg(feature = "esp32s2")]
+// We cannot just use the ROM function since it needs to allocate memory
+#[no_mangle]
+unsafe extern "C" fn strdup(str: *const i8) -> *const u8 {
+    trace!("strdup {:?}", str);
+
+    unsafe {
+        let s = core::ffi::CStr::from_ptr(str);
+        let s = s.to_str().unwrap();
+
+        let p = malloc(s.len() + 1);
+        core::ptr::copy_nonoverlapping(str, p as *mut i8, s.len() + 1);
+        p as *const u8
+    }
+}
+
+// We cannot just use the ROM function since it calls `__getreent`
+//
+// From docs: The __getreent() function returns a per-task pointer to struct
```

Comment:
I was looking into why on earth `atoi` needed to be reentrant, it seems it can set `errno` and therefore needs a thread local `errno` to set to avoid race conditions - annoying.